tests/std/tests/GH_002989_nothrow_unwrappable/test.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <filesystem>
+#include <list>
+#include <type_traits>
+#include <vector>
+#include <xutility>
+
+#if _HAS_CXX20
+#include <ranges>
+#endif
+
+using namespace std;
+using filesystem::path;
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+struct Predicate {
+    template <class T>
+    bool operator()(const T&) const {
+        return true;
+    }
+};
+
+template <class It, bool CopyUnwrapNothrow = true>
+void do_single_test() {
+    STATIC_ASSERT(_Is_nothrow_unwrappable_v<It>);
+    STATIC_ASSERT(_Is_nothrow_unwrappable_v<It&&>);
+    STATIC_ASSERT(noexcept(_Get_unwrapped(declval<It>())));
+
+    STATIC_ASSERT(_Is_nothrow_unwrappable_v<const It&> == CopyUnwrapNothrow);
+    STATIC_ASSERT(noexcept(_Get_unwrapped(declval<const It&>())) == CopyUnwrapNothrow);
+    STATIC_ASSERT(_Is_nothrow_unwrappable_v<const It&&> == CopyUnwrapNothrow);
+    STATIC_ASSERT(noexcept(_Get_unwrapped(declval<const It&&>())) == CopyUnwrapNothrow);
+}
+
+template <class It, bool CopyUnwrapNothrow = true>
+void do_full_test() {
+    do_single_test<It, CopyUnwrapNothrow>();
+    do_single_test<reverse_iterator<It>, CopyUnwrapNothrow>();
+    do_single_test<move_iterator<It>, CopyUnwrapNothrow>();
+
+#ifdef __cpp_lib_concepts // TRANSITION, GH-395
+    using R = ranges::subrange<It, It>;
+
+    do_single_test<ranges::iterator_t<R>, CopyUnwrapNothrow>();
+
+    // TRANSITION, GH-2997
+    do_single_test<ranges::iterator_t<ranges::filter_view<R, Predicate>>, true>();
+    // TRANSITION, GH-2997
+    do_single_test<ranges::iterator_t<ranges::transform_view<R, Predicate>>, true>();
+    if constexpr (ranges::bidirectional_range<R>) {

```

Comment:
You could spell this more concisely as:
```suggestion
    if constexpr (bidirectional_iterator<It>) {
```